### ISSUE TITLE
Fully differentiable training doesn't work with lower batch size

### ISSUE THREAD

**Issue by alessiobonfiglio** [aabbas90/COPS#1]
I don't have available any gpu with enough memory to execute your training scripts with the config you provided.
Because of that I have executed the pretrain on cityscapes with a batchsize of 2 (instead of 12), ending with a network with the following validation metrics (note that not having the small small validation set, I simply used the whole cityscapes validation set):
|PQ|SQ|RQ|PQ_th|SQ_th|RQ_th|PQ_st|SQ_st|RQ_st|
|----|----|-----|--------|---------|--------|---------|--------|--------|
|54.8540|80.6902|66.7585|42.9238|79.5514|53.7625|63.5305|81.5184|76.2101|

(the training total_loss was 1.501 at the final iteration)
I then used the last checkpoint of this pretrain for the fully differentiable training, with a bachsize of 8 (instead of 24), but during the training the total_loss went up to around 9.3, while the pq and sq went down to 0 (both the training and the evaluation ones at the end). I also tried lowering the learning rate, but the results were always the same.

Is this a known issue? Are there any solutions?

PS: in the paper you said you used {1, 4, 16, 32, 64, 128} as edge distances with cityscapes, but in your training configuration the maximum distance is 64. Is it intentional?

**Comment by aabbas90** [comment]
Hi,

Thanks and sorry for late reply. 

1. Unfortunately, this issue is because Panoptic Quality is not separable w.r.t expectation so if some class is present very less in the batch (e.g. only one person instance in the batch size of 8) and even if it has some small misclassification then it would give a huge gradient value in the loss. If you cannot afford large batch size then perhaps you can use some kind of [model checkpointing](https://pytorch.org/docs/stable/checkpoint.html) to save GPU memory and still run batch size 24 instead of 8? Another option might be to apply gradient clipping [here](https://github.com/aabbas90/COPS/blob/327d404d197baae6b7c3d35580bcd0b017dd0571/affinityNet/panoptic_affinity/losses.py#L136). Let me know if this still does not help.

2. Thanks for catching the edge distances issue. You are right the edge distance are indeed inconsistent. Btw the distance of 1 corresponds to 4 in full resolution i.e. these distances are computed w.r.t 4x downsampled output.
